src/files/DCDFile.cpp

```diff
@@ -0,0 +1,457 @@
+      fileStream.write((char*)&atomsNumBlockSize, sizeof(unsigned int));
+
+      /* If some atoms of the MD or MC simulation are frozen (i.e. never moving ) it is useless to store their coordinates more than once.
+       * In that case a list of Free atoms (moving ones) is written at the end of the header part of the dcd.
+       */
+      if (numFreeAtoms != numAtoms)
+      {
+         unsigned int freeAtomsBlockSize = numFreeAtoms * sizeof(int);
+         fileStream.write((char*)&freeAtomsBlockSize, sizeof(unsigned int));
+         fileStream.write((char*)freeAtoms, sizeof(int) * numFreeAtoms);
+         fileStream.write((char*)&freeAtomsBlockSize, sizeof(unsigned int));
+      }
+   }
+}
+
+void DCDFile::readFrame(std::vector<float>& X, std::vector<float>& Y,
+   std::vector<float>& Z, int toFrame)
+{
+   //dcd files store all the atoms in the first frame (even if there are some frozen atoms)
+   //and only store free atoms in the other frames
+   int frameAtmNum;
+   if (ICNTRL[8] != 0 && frameNumber != 0)
+   {
+      frameAtmNum = numAtoms - ICNTRL[8];
+   }
+   else
+   {
+      frameAtmNum = numAtoms;
+   }
+
+   X.reserve(frameAtmNum);
+   Y.reserve(frameAtmNum);
+   Z.reserve(frameAtmNum);
+
+   unsigned int fortcheck1, fortcheck2;
+
+   float* tmpX = new float[frameAtmNum];
+   float* tmpY = new float[frameAtmNum];
+   float* tmpZ = new float[frameAtmNum];
+
+   if (ICNTRL[10])
+   {
+      fileStream.read((char*)&fortcheck1, sizeof(unsigned int));
+      fileStream.read((char*)pbc, sizeof(double) * 6);
+      fileStream.read((char*)&fortcheck2, sizeof(unsigned int));
+      checkFortranIOerror(__FILE__, __LINE__, fortcheck1, fortcheck2);
+   }
+
+   // X
+   fileStream.read((char*)&fortcheck1, sizeof(unsigned int));
+   fileStream.read((char*)tmpX, sizeof(float) * frameAtmNum);
+   fileStream.read((char*)&fortcheck2, sizeof(unsigned int));
+   checkFortranIOerror(__FILE__, __LINE__, fortcheck1, fortcheck2);
+
+   // Y
+   fileStream.read((char*)&fortcheck1, sizeof(unsigned int));
+   fileStream.read((char*)tmpY, sizeof(float) * frameAtmNum);
+   fileStream.read((char*)&fortcheck2, sizeof(unsigned int));
+   checkFortranIOerror(__FILE__, __LINE__, fortcheck1, fortcheck2);
+
+   // Z
+   fileStream.read((char*)&fortcheck1, sizeof(unsigned int));
+   fileStream.read((char*)tmpZ, sizeof(float) * frameAtmNum);
+   fileStream.read((char*)&fortcheck2, sizeof(unsigned int));
+   checkFortranIOerror(__FILE__, __LINE__, fortcheck1, fortcheck2);
+
+   for (int it = 0; it < frameAtmNum; it++)
+   {
+      X.emplace_back(tmpX[it]);
+      Y.emplace_back(tmpY[it]);
+      Z.emplace_back(tmpZ[it]);
+   }
+
+   delete[] tmpX;
+   delete[] tmpY;
+   delete[] tmpZ;
+   frameNumber++;
+}
+
+void DCDFile::writeFrame(std::vector<float>& x, std::vector<float>& y,
+   std::vector<float>& z)
+{
+   if (fileStream.is_open())
+   {
+      assert((x.size() == y.size()) && (y.size() == z.size()));
+
+      float* tmpX = new float[x.size()];
```

Comment:
Same as above, allocating extra memory is not needed here

Reply:
Right!